### DISCUSSION TITLE
Test still passing if I execute vm.prank(address(1)) right before withdrawing the funds. Project: foundry-fund-me-f23

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#289] [Q&A] [answered]
I modified the `testUserCanFundInteraction` test by adding a `vm.prank(address(1))` right before withdrawing funds. This should change the user withdrawing the funds, to a new address, who is NOT the owner of the `FundMe.sol` contract. This should fail the test as in the `withdraw() public onlyOwner` function of Fundme.sol contract, we have an `onlyOwner()` modifier ensuring the owner can only be the one withdrawing.

Here is my `testUserCanFundInteraction` test:
```solidity
function testUserCanFundInteraction() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        vm.prank(USER); // introducing other user, hence test should fail
        vm.deal(USER, STARTING_BALANCE);

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));

        assertEq(address(fundMe).balance, 0);
    }
```
I'm wondering how the test passed when it was deployed by a `DeployFundMe.s.sol`script hence the address of `DeployFundMe.s.sol`script become the owner of `Fundme.sol`. Now when a prank user is withdrawing the funds, how did it pass?

An explanation will help. Thanks!

**Comment by Naman1729** [comment, marked as answer]
Hey @alfheimrShiven, vm.prank(User) will have its effect only on the line:

```solidity
WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
``` 

In your WithdrawFundMe function we have put the vm.startBroadcast() which use the deployer private key to call the withdraw() function of FundMe contract and also remember that Broadcast and prank does not work simultaneously . if you put vm.prank() just before this:

```solidity
withdrawFundMe.withdrawFundMe(address(fundMe));
```

then you will get an error .

**Reply by alfheimrShiven** [reply]
Thanks for explaining this @Naman1729 . Got to learn a few things about ```vm.prank()```. Cheers!

**Comment by ghost** [comment]
This is my personal opinion: I wouldn't include this as part of my interaction tests because I see it as an *overkill*.

This was tested as part of our unit tests but you could try a staging test but not on `Interaction.s.sol` but on the `FundMe` contract to verify the behaviour on a testnet.

The interaction scripts are meant to interact with an already deployed contract, so in my opinion you should be testing if the script can make the transaction and that's it, not testing if it reverts for X, Y or Z reason.

**Reply by alfheimrShiven** [reply]
Hey @pacelliv , I was just curious hence tried it. Got to learn a few things from this experiment, like:

1. ```vm.startBroadcast()``` and ```vm.prank(USER)``` cannot be used simultaneously as pointed out by @Naman1729 .
2. ```vm.startBroadcast()``` uses the deployer as the user, hence should be used for important transactions like deployments and transfers.
3. ```vm.prank()``` only affects the immediate next call and not the scope. Hence it doesn't change the actual caller of the function.

_I will not be including the prank as a permanent part of my source code. Thanks._
